### PR TITLE
Initial 1.0 implementation of Archipelago patch support

### PR DESCRIPTION
**Pull request by Ars-Ignis**
This is the initial implementation of Archipelago support for Crystalis Randomizer.  This is the basic approach taken by this implementation:

1. The Crystalis APWorld knows about all the flags and logic, including being able to make
   any random decisions that impact the logic directly. It only decides as much as it needs
   to in order to give proper logic to the Archipelago Generator process.
2. After fill is complete, the APWorld uses the generate_output step to write all of the
   relevant information into a .json file (including flagstring, seed, and any random 
   decisions made by the generator for logic), which then gets compressed and stored as an
   .apcrys file (secretly a .zip file with a bespoke file extension.
3. Once the multiworld is generated, the host will give the .apcrys file to the player, and
   the player will take the .apcrys file and their vanilla ROM to our website to apply the 
   specific data from the patch file instead of making random decisions. Any flags that
   cause randomness and don't affect logic get decided here.
4. In addition, using an .apcrys file in this way will add a bunch of custom ASM, mainly
   focused in a) separating item flags from location flags, and b) communicating with the
   Archipelago client as appropriate.
5. The user can now use this patched ROM in BizHawk, along with Archipelago's BizHawkClient
   to connect to their multiworld and start playing.
   
This PR/branch of the Crystalis Randomizer code concerns itself with steps 3 and 4 of the above process.  For details on the APWorld implementation (covering steps 1, 2, and 5), see [this repository](https://github.com/Ars-Ignis/Archipelago/tree/crystalis).  For additional thoughts on why I decided to take this approach (as well as some future work), see notes/archipelago.

### CODE REVIEW

File: notes/archipelago [hunk 1]
```diff
@@ -60,3 +60,86 @@ potentially be very efficient about defragging early on?
...
+	-Gifting interface? (maybe take over pawn-shops? not convinced this makes sense, but I'm willing to consider it)
+	-Option for local items being baked into the ROM? (Probably not, given how item flags end up working, but you never know)
+	
+Other notes:
+  Currently we don't bake any information into the ROM about what items are. However, this information
+  is available from the Archipelago generation. I am resistant to putting it into the patch data, as it
+  would result in the patch data doubling as a personal spoiler log for anyone with a little technical 
+  savvy. The might be a world where only local items get baked into the ROM, in which case I can just use
+  item IDs, which would do a sufficient amount of obfuscation, hopefully. I've also toyed around with the
+  idea of making the Crystalis Randomizer code connect to the Archipelago server as a client; then all the
```
Comment (shicks) [inline notes/archipelago:140]: Likely impossible with our current JS-only hosting situation, due to CSP or other sorts of security restrictions.  My guess is we'd require at least some sort of relay server to do the communication.

Comment (Ars-Ignis) [reply notes/archipelago:140]: That's fair.  Honestly it's kind of a janky idea to begin with, so probably not worth exploring in depth at this time.

File: src/js/patch.ts [hunk 2]
```diff
@@ -628,9 +651,17 @@ function randomizeWalls(rom: Rom, flags: FlagSet, random: Random): void {
...
     partition.get(location.data.area).push(location);
   }
-  for (const locations of partition.values()) {
+  for (const entry of partition.entries()) {
+    const area = entry[0];
+    const locations = entry[1];
     // pick a random wall type.
-    const elt = random.nextInt(4);
+    let elt: number = 0;
+    if (predetermined && predetermined.wallMap.has(area.name)) {
```
Comment (shicks) [inline src/js/patch.ts:659]: `predetermined?.wallMap.has(...)`

File: src/js/patch.ts [hunk 3]
```diff
@@ -281,13 +289,13 @@ async function shuffleInternal(rom: Uint8Array,
 
   // NOTE: Shuffle mimics and monsters *after* shuffling maps, but before logic.
   if (flags.shuffleMimics()) shuffleMimics(parsed, flags, random);
-  if (flags.shuffleMonsters()) shuffleMonsters(parsed, flags, random);
+  if (flags.shuffleMonsters()) shuffleMonsters(parsed, flags, random, (predetermined !== undefined && predetermined.fromArchipelago));
```
Comment (shicks) [inline src/js/patch.ts:292]: `predetermined?.fromArchipelago`

File: src/js/patch.ts [hunk 4]
```diff
@@ -101,7 +101,7 @@ function defines(flags: FlagSet,
     _ALLOW_TELEPORT_OUT_OF_TOWER: true,
     _AUDIBLE_WALLS: flags.audibleWallCues(pass),
     _AUTO_EQUIP_BRACELET: flags.autoEquipBracelet(pass),
-    _ARCHIPELAGO: predetermined != undefined && predetermined.fromArchipelago,
+    _ARCHIPELAGO: predetermined?.fromArchipelago == true,
```
Comment (shicks) [inline src/js/patch.ts:104]: This is fine to be undefined, so you don't need to do `== true`.

**Pull request merged**